modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -272,10 +272,21 @@ private void extractFromValues(
         Mappings.TargetMapping mapping = null;
 
         if (projections != null) {
-            projections = replaceInputRefs(projections);
+            List<RexNode> projectionsReplaced = replaceInputRefs(projections);
 
-            mapping = RexUtils.inversePermutation(projections,
-                    table.getRowType(Commons.typeFactory()), true);
+            boolean refFound = !projectionsReplaced.equals(projections);
+
+            if (refFound) {
```

Comment:
this is not binary choice (all refs vs all expressions). Add the following test case t see the problem `TEST("t(C1, C2) SELECT 100, t.x FROM (SELECT 100) as T(x)", TABLE_C1_NULLABLE_C2)`

Reply:
thanks ! i radically refactor the approach, can you check it for now ?